source/common/http/http1/codec_impl.cc

```diff
@@ -411,8 +411,9 @@ int ServerConnectionImpl::onHeadersComplete(HeaderMapImplPtr&& headers) {
         http_parser_pause(&parser_, 1);
       }
 
-    } else if (parser_.method == http_method::HTTP_POST ||
-               parser_.method == http_method::HTTP_PUT) {
+    } else if ((parser_.method == http_method::HTTP_POST ||
+                parser_.method == http_method::HTTP_PUT) &&
+               headers->ContentLength() == nullptr) {
```

Comment:
nit: I think the following comment is a bit misleading. It should be something like: "if content-length is not specified for certain methods..." Can we also reference new RFC section which says that we should do this? The rules have changed slightly at least according to new RFC and I'm not even completely sure we should do this change. (It's probably OK for HTTP/1.1, but will it break anyone else?) cc @jpinner-lyft